### PR TITLE
Space - Suely - Chelsea

### PR DESCRIPTION
**Pull request by ChelseaC13**
# Assignment Submission: OO Ride Share

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
How did getting up to speed on the existing codebase go? If it went well, what worked? If it didn't go well, what will you do differently next time? | It was hard in the beginning. Figuring out what definitions were calling step by step helped. Going line by line and figuring out what is being asked was key.
What _inheritance_ relations exist between classes? | there were multiple children to the parent CSV_record class with the exceptions of Trip_dispatcher.
What _composition_ relations exist between classes? | We were able to replicate functions by just calling functions in the module. Like a Rideshare.
Describe a decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | When is the best place/time to make a helper method? To solve this we just thought,  would we need this function somewhere else? If yes then making a helper method would be a good idea
Give an example of a _template method_ that you implemented for this assignment | We used self.from_csv(record) as a template method.
Give an example of a _nominal test_ that you wrote for this assignment | Returns the total driver revenue and making sure it is numeric would be nominal.
Give an example of an _edge case test_ that you wrote for this assignment | An edge case would be revenue for a trip below the trip fee. Example checking that a trip that costs $1.60 has no fee taken.
What is a concept that you gained more clarity on as you worked on this assignment | testing!

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    def initialize(id:, name:, vin:, status: :AVAILABLE, trips: nil)
+      super(id)
+
+      @name = name
+      @vin = vin
+      @status = status
+      @trips = trips || []
+
+      unless @vin.length == 17 && (@vin.is_a? String)
+        raise ArgumentError.new ('Your VIN is wrong.')
```
Comment (kaidamasaki) [inline lib/driver.rb:20]: Including the bad arguments in your `ArgumentError` messages is helpful (also avoid spaces before argument lists):

```suggestion
        raise ArgumentError.new('Your VIN (#{vin}) is wrong.')
```

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      end
+
+      approved_status = [:AVAILABLE, :UNAVAILABLE]
+
+      unless approved_status.include?(@status)
+        raise ArgumentError.new ('You must provide one of the following statuses :available, :unavailable')
```
Comment (kaidamasaki) [inline lib/driver.rb:26]: Symbols are case sensitive:

```suggestion
        raise ArgumentError.new ('You must provide one of the following statuses :AVAILABLE, :UNAVAILABLE')
```

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      end
+    end
+
+    def add_trip(trip)
+      @trips << trip
+    end
+
+    def average_rating
+      completed = @trips.select { |trip| trip.rating != nil }
+      completed.map {|trip| trip.rating}.sum.to_f / (completed.size == 0 ? 1 : completed.size)
```
Comment (kaidamasaki) [inline lib/driver.rb:36]: You can use `sum` with a block to replace a usage of `map`.  Good use of a ternary though.  😄 

```suggestion
      completed.sum {|trip| trip.rating}.to_f / (completed.size == 0 ? 1 : completed.size)
```

File: lib/trip.rb [hunk 4]
```diff
@@ -4,63 +4,89 @@
...
+      elsif driver_id
+        @driver_id = driver_id
+      else
+        raise ArgumentError, 'Driver or driver_id is required'
+      end
 
-      @start_time = start_time
+      if start_time.class == String
+        start_time = Time.parse(start_time)
+      end
```
Comment (kaidamasaki) [inline lib/trip.rb:42]: Since you're parsing times in `from_csv` you don't need to do that defensively here.

```suggestion
```

File: lib/passenger.rb [hunk 5]
```diff
@@ -16,6 +16,16 @@ def add_trip(trip)
       @trips << trip
     end
 
+    def net_expenditures
+      completed = @trips.select { |trip| trip.cost != nil }
```
Comment (kaidamasaki) [inline lib/passenger.rb:20]: Nice use of `select`.  😃

File: lib/trip_dispatcher.rb [hunk 6]
```diff
@@ -9,33 +9,94 @@ class TripDispatcher
...
               #{trips.count} trips, \
               #{drivers.count} drivers, \
               #{passengers.count} passengers>"
     end
 
+    def request_trip(passenger_id)
+      passenger = find_passenger(passenger_id)
+      raise ArgumentError, "Passenger id not found" if passenger == nil
+
+      driver = @drivers.select {|driver| driver.status == :AVAILABLE}.first
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb:39]: You can use `find` to make this a little clearer:

```suggestion
      driver = @drivers.find {|driver| driver.status == :AVAILABLE}
```

File: lib/trip_dispatcher.rb [hunk 7]
```diff
@@ -9,33 +9,94 @@ class TripDispatcher
...
+
+    # Helper Method to Find In Progress Trips in an Array of Trips
+    def in_progress(trips)
+      return trips.find { |trip| trip.end_time == nil } # nil if no in progress trip, otherwise that trip
+    end
+
+    # Helper Method to Find the End Time of the Most Recent Trip in an Array of Trips
+    def recent_end_time(trips)
+      return trips.sort_by { |trip| trip.end_time }.last.end_time 
+    end
```
Comment (kaidamasaki) [inline lib/trip_dispatcher.rb:87]: I would expect both of these to use `@trips` instead of taking `trips` as a parameter.

```suggestion
    def in_progress
      return @trips.find { |trip| trip.end_time == nil } # nil if no in progress trip, otherwise that trip
    end

    # Helper Method to Find the End Time of the Most Recent Trip in an Array of Trips
    def recent_end_time
      return @trips.sort_by { |trip| trip.end_time }.last.end_time 
    end
```

File: test/driver_test.rb [hunk 8]
```diff
@@ -128,9 +132,101 @@
 
       expect(@driver.average_rating).must_be_close_to (5.0 + 1.0) / 2.0, 0.01
     end
+
+    it "average rating works with in-progress trips" do
+
+      @driver = RideShare::Driver.new(
```
Comment (kaidamasaki) [inline test/driver_test.rb:138]: You should avoid reassigning instance variables in tests, it can cause some confusing bugs.

Instead create a local variable and reference it below:
```suggestion
      driver = RideShare::Driver.new(
```

File: test/passenger_test.rb [hunk 9]
```diff
@@ -69,6 +77,144 @@
...
+        rating: 5
+        )
+
+      @trip3 = RideShare::Trip.new(
+        id: 11,
+        driver: driver,
+        passenger: @passenger,
+        start_time: Time.new(2016, 8, 11) 
+        )
+      end
```
Comment (kaidamasaki) [inline test/passenger_test.rb:122]: Incorrect indentation:

```suggestion
    before do
      
      @passenger = RideShare::Passenger.new(
        id: 9,
        name: "Merl Glover III",
        phone_number: "1-602-620-2330 x3723",
        trips: []
      )

      driver = RideShare::Driver.new(
        id: 54,
        name: "Rogers Bartell IV",
        vin: "1C9EVBRM0YBC564DZ",
        status: :AVAILABLE      
      )
  
      @trip1 = RideShare::Trip.new(
        id: 8,
        driver: driver,
        passenger: @passenger,
        start_time: Time.new(2016, 8, 8),
        end_time: Time.new(2016, 8, 9),
        cost: 10,
        rating: 5
      )

      @trip2 = RideShare::Trip.new(
        id: 10,
        driver: driver,
        passenger: @passenger,
        start_time: Time.new(2016, 8, 11),
        end_time: Time.new(2016, 8, 12),
        cost: 40,
        rating: 5
      )

      @trip3 = RideShare::Trip.new(
        id: 11,
        driver: driver,
        passenger: @passenger,
        start_time: Time.new(2016, 8, 11) 
      )
    end
```

File: test/passenger_test.rb [hunk 10]
```diff
@@ -69,6 +77,144 @@
...
+        )
+
+        @trip3 = RideShare::Trip.new(
+        id: 11,
+        driver: driver,
+        passenger: @passenger,
+        start_time: Time.new(2016, 8, 11)
+        )
+
+      end
```
Comment (kaidamasaki) [inline test/passenger_test.rb:184]: Incorrect indentation:

```suggestion
    before do
      @passenger = RideShare::Passenger.new(
        id: 9,
        name: "Merl Glover III",
        phone_number: "1-602-620-2330 x3723",
        trips: []
      )

      driver = RideShare::Driver.new(
        id: 54,
        name: "Rogers Bartell IV",
        vin: "1C9EVBRM0YBC564DZ",
        status: :AVAILABLE      
      )
        
      @trip1 = RideShare::Trip.new(
        id: 8,
        driver: driver,
        passenger: @passenger,
        start_time: Time.new(2016, 8, 8),
        end_time: Time.new(2016, 8, 9),
        cost: 10,
        rating: 5
      )

      @trip2 = RideShare::Trip.new(
        id: 10,
        driver: driver,
        passenger: @passenger,
        start_time: Time.new(2016, 8, 11),
        end_time: Time.new(2016, 8, 12),
        cost: 40,
        rating: 5
      )

      @trip3 = RideShare::Trip.new(
        id: 11,
        driver: driver,
        passenger: @passenger,
        start_time: Time.new(2016, 8, 11)
      )

    end
```

File: test/passenger_test.rb [hunk 11]
```diff
@@ -69,6 +77,144 @@
...
+
+    it "Returns the total time the passenger has had on trips." do
+      expect(@passenger.total_time_spent).must_equal 0
+      expect(@passenger.total_time_spent).must_be_kind_of Numeric
```
Comment (kaidamasaki) [inline test/passenger_test.rb:188]: These assertions are in the wrong order.  If `total_time_spent` is `0` then it must already be `Numeric`:

```suggestion
      expect(@passenger.total_time_spent).must_be_kind_of Numeric
      expect(@passenger.total_time_spent).must_equal 0
```

File: test/trip_test.rb [hunk 12]
```diff
@@ -41,5 +44,44 @@
...
+
+      @trip_data[:start_time] = @trip_data[:end_time] + 1 
+      expect do
+        RideShare::Trip.new(@trip_data)
+      end.must_raise ArgumentError
+    end
+
+    it "calculates the duration of the trip corectly" do
+      expect(@trip.duration).must_be_kind_of Float
+      expect(@trip.duration).must_equal 1500.0
```
Comment (kaidamasaki) [inline test/trip_test.rb:62]: You should avoid using `must_equal` with `Float`s because they are imprecise.  You should use `must_be_close_to` instead.

This requires that they be within `0.1`:
```suggestion
      expect(@trip.duration).must_be_close_to 1500.0, 0.1
```

File: lib/driver.rb [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    end
+
+    def total_revenue
+      fee = 1.65
+      completed = @trips.select { |trip| trip.cost != nil }
+      return completed.map {|trip| trip.cost < fee ? 0 : trip.cost - fee}.sum * 0.8
```
Comment (kaidamasaki) [inline lib/driver.rb:42]: You can also write this using `reject` (which might read a little clearer).

```suggestion
      return completed.reject {|trip| trip.cost < fee}.sum { |trip| trip.cost - fee} * 0.8
```

**Comment by kaidamasaki** [comment]
# OO Ride Share



## Major Learning Goals/Code Review



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
The code demonstrates individual learning about `Time` and the responsibility of `Trip.from_csv`, and uses `Time.parse` in `Trip.from_csv` | ✔️
The code demonstrates breaking out complex logic in helper methods, such as making a helper method in `Trip` to calculate duration | ✔️
There are tests for the nominal cases for the `Passenger#net_expenditures` and `Passenger#total_time_spent` | ✔️
There is at least one edge case test for either `Passenger#net_expenditures` or `Passenger#total_time_spent` testing if the passenger has no trips | ✔️
Practices inheritance. `Driver` inherits from `CsvRecord`, and implements `from_csv` | ✔️
Employs problem-solving and implements `Driver#average_rating` and `Driver#total_revenue` | ✔️
Implements the `TripDispatcher#request_trip`, which creates an instance of `Trip` with a driver and passenger, adds the new trip to `@trips`, and changes the status of the driver | ✔️
Practices composition. In `TripDispatcher#request_trip`, the driver gets connected to the new trip, the passenger gets connected to the new trip | ✔️
Practices git with at least 10 small commits and meaningful commit messages | ✔️ (Some of your commit messages were a bit vague though, for example: "trying again" and "fixing errors".  What are you trying again?  What errors did you fix?)

## Testing Requirements

| Testing Requirement | yes/no |
| --- | --- |
There is reasonable test coverage for wave 1, and all wave 1 tests pass | ✔️
There is reasonable test coverage for wave 2, and all wave 2 tests pass | ✔️
Wave 3: Tests in wave 1 and wave 2 explicitly test that only completed trips should be calculated (and ignore in-progress trips) | ✔️
There is reasonable test coverage for `TripDispatcher#request_trip`, and all tests pass | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 3+ in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 6+ in Code Review && 2+ in Functional Requirements |
| Red (Not at Standard) | 0-5 in Code Review or 0,1 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | (See comments.)
| Elegant/Clever |
| Descriptive/Readable | ✅
| Concise |
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Good job!  Here are some small things you can do to clean up your code.

**Review by kaidamasaki — COMMENTED** [review]
Good job!  Here are a somet things you can do to clean up your code.
